### ISSUE TITLE
require-await rule should allow async keyword for function returning the result of an async function

### ISSUE THREAD

**Issue by pvdlg** [eslint/eslint#10829]
**What rule do you want to change?**
[require-await](https://eslint.org/docs/rules/require-await)

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
Default behavior.

Maybe add an option that would enforce using the `async` keyword for function that return the result of calling another `async` function. That would make sure async function are always clearly identified with the `async` keyword. Or instead of an option in this rule it might be another rule that enforce using the `async` keyword for Promise returning functions.

**Please provide some example code that this change will affect:**

```js
async doSomething() {
  await doSomethingElse();
}

async function foo() {
    return doSomething();
}

async function bar() {
    const result = await doSomething();
    return result;
}
```

**What does the rule currently do for this code?**
The rule return an error saying `foo` should not use the `async` keyword.
There is no error for `bar` even though they are functionally identical.

**What will the rule do after it's changed?**
The code should be valid as `foo` is actually an `async` function as it returns a `Promise` (the result of calling `doSomething` which is an `async` function).

It also doesn't make sense to return an error for `foo` but not for `bar`.

**Comment by kaicataldo** [comment]
Hi - apologies for the delay in response.

Given the limitations of static analysis and the fact that ESLint only runs on one file at a time, I don't know how ESLint could check this in a reliable fashion. Do you have any ideas?

**Comment by pvdlg** [comment]
Hi! Yes we would need to know if a function is declared as `async` or return a `Promise`, and that function might be declared in another file.

Maybe in that case the rule could resolve and parse the file that declares the imported function? Or is there any plan in the future for eslint to create the AST fro the entire project?

**Comment by kaicataldo** [comment]
> Maybe in that case the rule could resolve and parse the file that declares the imported function? 

ESLint currently doesn't provide any APIs to allow for this, but it could potentially be done in a plugin ([eslint-plugin-import](https://www.npmjs.com/package/eslint-plugin-import) will check for the existence of modules, for instance).

> Or is there any plan in the future for eslint to create the AST fro the entire project?

Not currently, no.

Even if we implemented the things above, we're still limited by the constraints of static analysis. This doesn't feel possible to me without the help of a typing system (such as TypeScript).

**Comment by chris-perts** [comment]
What about just an option on the rule, like `allowImplicitAsyncReturn` that doesn't throw an error for code like this (when `y` is async):

`const getIt = async x => y()`

ESLint would not be expected to know that `y` is an async function, it would just silence the error in this case because you had the option set.

As it is, when I turn on both `no-return-await` and `require-await`, there's no implicit return version of this function that doesn't raise an error, which seems silly. Specifically, `const getIt = async x => await y()` violates `no-return-await`, while `const getIt = async x => y()` violates `require-await`.

I think it's reasonable to want all three things, 1) to have a reminder that (most of the time) an async function needs `await` somewhere, 2) to have a reminder that `return await` is redundant, and 3) to use terse arrow functions.

**Comment by chris-perts** [comment]
@kaicataldo I'll take your suggestion of just dropping `async`, but I don't love that either, since it's nice for documentation to label async functions as such. It's still the best available option though. Thanks.
